File: app/src/main/java/com/fiatlife/app/domain/model/Budget.kt
package com.fiatlife.app.domain.model

import kotlinx.serialization.Serializable
import java.util.Calendar

/**
 * Monthly budget model — the Android counterpart of the web `BudgetConfig`.
 *
 * A budget is a single per-user config record (d_tag `fiatlife/budget`), like
 * salary. It holds a monthly target per spending category plus, for categories
 * that aren't backed by recurring bills, a manually-entered "spent so far this
 * month" figure.
 *
 * Actuals are hybrid:
 *  - [BudgetCategoryKind.BILL] categories pull their actual from existing bill
 *    data (sum of monthly-equivalent amounts per [BillGeneralCategory]).
 *  - [BudgetCategoryKind.VARIABLE] categories use the user-entered [CategoryBudget.manualSpent].
 *
 * `manualSpent` resets at the start of each calendar month (see
 * [rollBudgetPeriod]); targets persist month to month.
 */

@Serializable
enum class BudgetCategoryKind { BILL, VARIABLE }

@Serializable
data class CategoryBudget(
    /** [BillGeneralCategory] name for BILL kind; a variable category key otherwise. */
    val key: String = "",
    val kind: BudgetCategoryKind = BudgetCategoryKind.VARIABLE,
    /** Monthly budget target in dollars. */
    val target: Double = 0.0,
    /** Manually-entered spend so far this month (variable categories only). */
    val manualSpent: Double = 0.0
)

@Serializable
data class BudgetConfig(
    val id: String = "",
    /** Calendar month the [CategoryBudget.manualSpent] figures apply to, as "YYYY-MM". */
    val periodMonth: String = "",
    val categoryBudgets: List<CategoryBudget> = emptyList(),
    val updatedAt: Long = 0L
)

/** A variable spending category — the "general purchases" that aren't bills. */
data class VariableBudgetCategory(val key: String, val label: String)

val VARIABLE_BUDGET_CATEGORIES: List<VariableBudgetCategory> = listOf(
    VariableBudgetCategory("GROCERIES", "Groceries"),
    VariableBudgetCategory("DINING", "Dining Out"),
    VariableBudgetCategory("TRANSPORTATION", "Transportation/Fuel"),
    VariableBudgetCategory("ENTERTAINMENT", "Entertainment"),
    VariableBudgetCategory("SHOPPING", "Shopping"),
    VariableBudgetCategory("PERSONAL_CARE", "Personal Care"),
    VariableBudgetCategory("MISC", "Miscellaneous")
)

private val VARIABLE_KEYS: Set<String> = VARIABLE_BUDGET_CATEGORIES.map { it.key }.toSet()
private val BILL_KEYS: Set<String> = BillGeneralCategory.entries.map { it.name }.toSet()

fun variableBudgetCategoryLabel(key: String): String =
    VARIABLE_BUDGET_CATEGORIES.firstOrNull { it.key == key }?.label ?: key

/** Current calendar month as "YYYY-MM" in local time. */
fun currentBudgetPeriodMonth(now: Long = System.currentTimeMillis()): String {
    val cal = Calendar.getInstance().apply { timeInMillis = now }
    val year = cal.get(Calendar.YEAR)
    val month = cal.get(Calendar.MONTH) + 1
    return "%04d-%02d".format(year, month)
}

fun defaultBudgetConfig(now: Long = System.currentTimeMillis()): BudgetConfig =
    BudgetConfig(periodMonth = currentBudgetPeriodMonth(now))

/**
 * Roll the budget into the current month if stale: targets carry over, manual
 * spend resets to 0. Returns the same object when no change is needed.
 */
fun rollBudgetPeriod(config: BudgetConfig, now: Long = System.currentTimeMillis()): BudgetConfig {
    val period = currentBudgetPeriodMonth(now)
    if (config.periodMonth == period) return config
    return config.copy(
        periodMonth = period,
        categoryBudgets = config.categoryBudgets.map { it.copy(manualSpent = 0.0) }
    )
}

fun getCategoryBudget(config: BudgetConfig, key: String): CategoryBudget? =
    config.categoryBudgets.firstOrNull { it.key == key }

private fun sanitizeAmount(value: Double): Double = if (value.isFinite() && value > 0) value else 0.0

/** Immutably upsert a single category's target/spent. */
fun setCategoryBudget(
    config: BudgetConfig,
    key: String,
    kind: BudgetCategoryKind,
    target: Double? = null,
    manualSpent: Double? = null
): BudgetConfig {
    val existing = getCategoryBudget(config, key)
    val next = CategoryBudget(
        key = key,
        kind = kind,
        target = sanitizeAmount(target ?: existing?.target ?: 0.0),
        manualSpent = if (kind == BudgetCategoryKind.VARIABLE) {
            sanitizeAmount(manualSpent ?: existing?.manualSpent ?: 0.0)
        } else 0.0
    )
    val others = config.categoryBudgets.filterNot { it.key == key }
    return config.copy(categoryBudgets = others + next)
}

/** Last-write-wins merge that preserves the stable id when the incoming copy lacks one. */
fun mergeBudgetConfigPreserveId(incoming: BudgetConfig, existing: BudgetConfig?): BudgetConfig {
    if (existing == null) return incoming
    return if (incoming.id.isEmpty() && existing.id.isNotEmpty()) {
        incoming.copy(id = existing.id)
    } else incoming
}

data class BudgetRow(
    val key: String,
    val label: String,
    val kind: BudgetCategoryKind,
    val target: Double,
    /** BILL: derived from bills. VARIABLE: manualSpent. */
    val actual: Double
) {
    val remaining: Double get() = target - actual
    /** 0–100+, where >100 means over budget. */
    val percentUsed: Double
        get() = when {
            target > 0 -> actual / target * 100.0
            actual > 0 -> 100.0
            else -> 0.0
        }
}

data class BudgetSummary(
    val billRows: List<BudgetRow> = emptyList(),
    val variableRows: List<BudgetRow> = emptyList(),
    val totalTarget: Double = 0.0,
    val totalActual: Double = 0.0,
    val totalBillActual: Double = 0.0,
    val totalVariableActual: Double = 0.0,
    val takeHome: Double = 0.0
) {
    /** take-home minus everything budgeted (targets). */
    val unbudgeted: Double get() = takeHome - totalTarget
    /** take-home minus everything actually spent/committed. */
    val remaining: Double get() = takeHome - totalActual
}

/**
 * Build the displayed budget. Bill rows are shown for every general category
 * that has a target set or has bills; variable rows are always shown.
 */
fun computeBudgetSummary(
    config: BudgetConfig,
    billCategoryTotals: Map<BillGeneralCategory, Double>,
    takeHome: Double
): BudgetSummary {
    val billRows = BillGeneralCategory.entries.mapNotNull { cat ->
        val billActual = billCategoryTotals[cat] ?: 0.0
        val target = getCategoryBudget(config, cat.name)?.target ?: 0.0
        if (billActual <= 0.0 && target <= 0.0) return@mapNotNull null
        BudgetRow(
            key = cat.name,
            label = cat.displayName,
            kind = BudgetCategoryKind.BILL,
            target = target,
            actual = billActual
        )
    }

    val variableRows = VARIABLE_BUDGET_CATEGORIES.map { (key, label) ->
        val entry = getCategoryBudget(config, key)
        BudgetRow(
            key = key,
            label = label,
            kind = BudgetCategoryKind.VARIABLE,
            target = entry?.target ?: 0.0,
            actual = entry?.manualSpent ?: 0.0
        )
    }

    val totalBillActual = billRows.sumOf { it.actual }
    val totalVariableActual = variableRows.sumOf { it.actual }
    val totalTarget = billRows.sumOf { it.target } + variableRows.sumOf { it.target }

    return BudgetSummary(
        billRows = billRows,
        variableRows = variableRows,
        totalTarget = totalTarget,
        totalActual = totalBillActual + totalVariableActual,
        totalBillActual = totalBillActual,
        totalVariableActual = totalVariableActual,
        takeHome = takeHome
    )
}

File: apps/web/web/src/lib/budget.ts
/**
 * Monthly budget model — mirrors the planned Android `BudgetConfig`.
 *
 * A budget is a single per-user config record (d_tag `fiatlife/budget`), like
 * salary. It holds a monthly target for each spending category plus, for
 * categories that aren't backed by recurring bills, a manually-entered
 * "spent so far this month" figure.
 *
 * Actuals are hybrid:
 *  - "bill"-kind categories pull their actual from existing bill data
 *    (sum of monthly-equivalent amounts per `BillGeneralCategory`).
 *  - "variable"-kind categories use the user-entered `manualSpent`.
 *
 * `manualSpent` resets at the start of each calendar month (see
 * `rollBudgetPeriod`); targets persist month to month.
 */
import {
  ALL_GENERAL_CATEGORIES,
  GENERAL_CATEGORY_LABELS,
  type BillGeneralCategory,
} from "./bill";

export const BUDGET_D_TAG = "fiatlife/budget";

export type BudgetCategoryKind = "bill" | "variable";

export type CategoryBudget = {
  /** `BillGeneralCategory` name for bill-kind; a `VariableCategoryKey` otherwise. */
  key: string;
  kind: BudgetCategoryKind;
  /** Monthly budget target in dollars. */
  target: number;
  /** Manually-entered spend so far this month (variable categories only). */
  manualSpent: number;
};

export type BudgetConfig = {
  id: string;
  /** Calendar month the `manualSpent` figures apply to, as "YYYY-MM". */
  periodMonth: string;
  categoryBudgets: CategoryBudget[];
  updatedAt: number;
};

/** Spending categories that aren't recurring bills — the "general purchases". */
export const VARIABLE_CATEGORIES: { key: string; label: string }[] = [
  { key: "GROCERIES", label: "Groceries" },
  { key: "DINING", label: "Dining Out" },
  { key: "TRANSPORTATION", label: "Transportation/Fuel" },
  { key: "ENTERTAINMENT", label: "Entertainment" },
  { key: "SHOPPING", label: "Shopping" },
  { key: "PERSONAL_CARE", label: "Personal Care" },
  { key: "MISC", label: "Miscellaneous" },
];

const VARIABLE_LABELS: Record<string, string> = Object.fromEntries(
  VARIABLE_CATEGORIES.map((c) => [c.key, c.label]),
);

export function variableCategoryLabel(key: string): string {
  return VARIABLE_LABELS[key] ?? key;
}

/** Current calendar month as "YYYY-MM" in local time. */
export function currentPeriodMonth(now = Date.now()): string {
  const d = new Date(now);
  const month = `${d.getMonth() + 1}`.padStart(2, "0");
  return `${d.getFullYear()}-${month}`;
}

export function defaultBudgetConfig(now = Date.now()): BudgetConfig {
  return {
    id: "",
    periodMonth: currentPeriodMonth(now),
    categoryBudgets: [],
    updatedAt: 0,
  };
}

function isVariableKey(key: string): boolean {
  return VARIABLE_CATEGORIES.some((c) => c.key === key);
}

function isBillKey(key: string): boolean {
  return (ALL_GENERAL_CATEGORIES as string[]).includes(key);
}

function sanitizeAmount(value: unknown): number {
  const n = typeof value === "number" ? value : Number(value);
  if (!Number.isFinite(n) || n < 0) return 0;
  return n;
}

export function normalizeBudgetConfig(config: BudgetConfig): BudgetConfig {
  const seen = new Set<string>();
  const categoryBudgets: CategoryBudget[] = [];
  for (const entry of config.categoryBudgets ?? []) {
    const key = entry.key;
    if (!key || seen.has(key)) continue;
    if (!isVariableKey(key) && !isBillKey(key)) continue;
    seen.add(key);
    const kind: BudgetCategoryKind = isVariableKey(key) ? "variable" : "bill";
    categoryBudgets.push({
      key,
      kind,
      target: sanitizeAmount(entry.target),
      manualSpent: kind === "variable" ? sanitizeAmount(entry.manualSpent) : 0,
    });
  }
  return {
    ...config,
    periodMonth: config.periodMonth || currentPeriodMonth(),
    categoryBudgets,
  };
}

export function parseBudgetRecord(plaintext: string): BudgetConfig | null {
  try {
    const raw = JSON.parse(plaintext) as Partial<BudgetConfig> & {
      deleted?: boolean;
    };
    if (!raw || typeof raw !== "object" || raw.deleted) return null;
    return normalizeBudgetConfig({
      id: typeof raw.id === "string" ? raw.id : "",
      periodMonth:
        typeof raw.periodMonth === "string" ? raw.periodMonth : currentPeriodMonth(),
      categoryBudgets: Array.isArray(raw.categoryBudgets)
        ? (raw.categoryBudgets as CategoryBudget[])
        : [],
      updatedAt: typeof raw.updatedAt === "number" ? raw.updatedAt : 0,
    });
  } catch {
    return null;
  }
}

export function serializeBudget(config: BudgetConfig): string {
  return JSON.stringify(normalizeBudgetConfig(config));
}

/** Stable string for change detection (excludes `updatedAt`). */
export function budgetFingerprint(config: BudgetConfig): string {
  const sorted = [...config.categoryBudgets]
    .map((c) => `${c.key}:${c.kind}:${c.target}:${c.manualSpent}`)
    .sort();
  return `${config.periodMonth}|${sorted.join(",")}`;
}

/**
 * Roll the budget into the current month if it's stale: targets carry over,
 * manually-entered spend resets to 0. Returns the same object when no change.
 */
export function rollBudgetPeriod(
  config: BudgetConfig,
  now = Date.now(),
): BudgetConfig {
  const period = currentPeriodMonth(now);
  if (config.periodMonth === period) return config;
  return {
    ...config,
    periodMonth: period,
    categoryBudgets: config.categoryBudgets.map((c) => ({
      ...c,
      manualSpent: 0,
    })),
  };
}

export function getCategoryBudget(
  config: BudgetConfig,
  key: string,
): CategoryBudget | undefined {
  return config.categoryBudgets.find((c) => c.key === key);
}

/** Immutably upsert a single category's target/spent. */
export function setCategoryBudget(
  config: BudgetConfig,
  key: string,
  kind: BudgetCategoryKind,
  patch: { target?: number; manualSpent?: number },
): BudgetConfig {
  const existing = getCategoryBudget(config, key);
  const next: CategoryBudget = {
    key,
    kind,
    target: sanitizeAmount(patch.target ?? existing?.target ?? 0),
    manualSpent:
      kind === "variable"
        ? sanitizeAmount(patch.manualSpent ?? existing?.manualSpent ?? 0)
        : 0,
  };
  const others = config.categoryBudgets.filter((c) => c.key !== key);
  return { ...config, categoryBudgets: [...others, next] };
}

export type BudgetRow = {
  key: string;
  label: string;
  kind: BudgetCategoryKind;
  target: number;
  /** Bill-kind: derived from bills. Variable-kind: `manualSpent`. */
  actual: number;
  remaining: number;
  /** 0–100+, where >100 means over budget. */
  percentUsed: number;
};

export type BudgetSummary = {
  billRows: BudgetRow[];
  variableRows: BudgetRow[];
  totalTarget: number;
  totalActual: number;
  totalBillActual: number;
  totalVariableActual: number;
  takeHome: number;
  /** take-home minus everything budgeted (targets). */
  unbudgeted: number;
  /** take-home minus everything actually spent/committed. */
  remaining: number;
};

function makeRow(
  key: string,
  label: string,
  kind: BudgetCategoryKind,
  target: number,
  actual: number,
): BudgetRow {
  return {
    key,
    label,
    kind,
    target,
    actual,
    remaining: target - actual,
    percentUsed: target > 0 ? (actual / target) * 100 : actual > 0 ? 100 : 0,
  };
}

/**
 * Build the displayed budget. Bill-kind rows are shown for every general
 * category that has a target set or has bills; variable rows are always shown.
 */
export function computeBudgetSummary(input: {
  config: BudgetConfig;
  billCategoryTotals: Partial<Record<BillGeneralCategory, number>>;
  takeHome: number;
}): BudgetSummary {
  const { config, billCategoryTotals, takeHome } = input;

  const billRows: BudgetRow[] = [];
  for (const cat of ALL_GENERAL_CATEGORIES) {
    const billActual = billCategoryTotals[cat] ?? 0;
    const target = getCategoryBudget(config, cat)?.target ?? 0;
    if (billActual <= 0 && target <= 0) continue;
    billRows.push(
      makeRow(cat, GENERAL_CATEGORY_LABELS[cat], "bill", target, billActual),
    );
  }

  const variableRows: BudgetRow[] = VARIABLE_CATEGORIES.map(({ key, label }) => {
    const entry = getCategoryBudget(config, key);
    return makeRow(key, label, "variable", entry?.target ?? 0, entry?.manualSpent ?? 0);
  });

  const totalBillActual = billRows.reduce((s, r) => s + r.actual, 0);
  const totalVariableActual = variableRows.reduce((s, r) => s + r.actual, 0);
  const totalActual = totalBillActual + totalVariableActual;
  const totalTarget =
    billRows.reduce((s, r) => s + r.target, 0) +
    variableRows.reduce((s, r) => s + r.target, 0);

  return {
    billRows,
    variableRows,
    totalTarget,
    totalActual,
    totalBillActual,
    totalVariableActual,
    takeHome,
    unbudgeted: takeHome - totalTarget,
    remaining: takeHome - totalActual,
  };
}
